pkg/tasks/runner.go

```diff
@@ -33,25 +34,78 @@ func RunAll(cfg *config.Config, proj *project.Project, ui *termui.UI) (success b
 	}
 
 	for _, task := range taskList {
-		ctx.ui.TaskHeader(task.name(), task.header())
+		if t, ok := task.(taskWithPreRunValidation); ok {
+			err = t.preRunValidation(ctx)
+			if err != nil {
+				ctx.ui.TaskError(err)
+				return false, nil
+			}
+		}
+	}
 
-		err = task.perform(ctx)
+	for _, task := range taskList {
+		ctx.ui.TaskHeader(task.name(), task.header())
+		err = runTask(ctx, task)
 		if err != nil {
 			ctx.ui.TaskError(err)
-			return false, nil
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+func runTask(ctx *context, task Task) (err error) {
+	if t, ok := task.(taskWithPerform); ok {
```

Comment:
can this fail?

Reply:
it doesn't *fail*. The task either fulfill the interface or it does not. The `ok` variable is set accordingly.

Reply:
Me and my golang skills though it was a simple casting. This naturally cannot failed cause we've "implemented" the interface. that is 👌